untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,493 @@
+> Further, $\lambda$ and CL gain most of their purpose at second hand from such systems, just as an isolated chassis has little purpose in itself.
+>
+>  -- Felice Cardone, J. Roger Hindley, History of Lambda-calculus and Combinatory Logic
+
+After this, admittedly, a very brief look at the untyped lambda calculus, you hopefully have a bit of a feel for how it works.
+The primary point of interest is, of course, not as much in the calculus itself, but in its applications, of which there are many.
+However, the untyped kind doesn't find much use beyond the computability theory.
+Still, it's necessary to understand the untyped lambda calculus to meaningfully discuss the typed variety (or, more accurately, varieties).
+
+So, let us summarize:
+
+- lambda calculus is a theoretical framework on which functional programming languages are built;
+- lambda-calculus has two kinds, untyped and typed;
+- the pure untyped lambda calculus only consists of variables, abstractions (i.e. univariate functions), and function applications;
+- we can represent any computable function in pure untyped lambda calculus;
+- various encodings can be used to represent booleans, naturals, tuples, etc.
+
+The last three points are somewhat profound if you think about it: we only defined a single operation (two if you count the lambda abstraction definition), and that got us to anything theoretically computable.
+There are other constructions of the universal computer: Turing and Post machines, the Markov algorithm, and general recursive functions.
+But, arguably, nothing beats the simplicity of lambda calculus.
+One has to appreciate the elegance at least.
+
+## Exercises
+
+1. We introduced the implementation of the boolean conjunction for Church-encoded boolean values (the `and` operation).
+    Implement disjunction (`or`), inversion (`not`), and exclusive disjunction (`xor`).
+
+2. We did introduce one particular implementation for the successor function for Church-encoded naturals, `succ`.
+    It is not the only possible implementation, however.
+    Suggest another possible implementation, which is not alpha-equivalent to the one we introduced.
+
+3. Define an exponentiation function for Church-encoded naturals.
+
+4. Define subtraction of Church-encoded naturals using the predecessor function `pred`.
+    For the case when the difference is undefined in naturals, i.e. when the subtrahend is bigger than minuend, feel free to return $c_0$ or $\Omega\;\Omega$.
+
+5. Define a function taking two Church-encoded naturals and returning `tru` if those are equal and `fls` otherwise.
+
+6. Implement a pure untyped lambda calculus interpreter in your favorite programming language.
+
+    One question you might have is how to handle name collisions and encode alpha-equivalence.
```

Comment:
I would hide this under some Tips for handling name collisions spoiler or something, it puts too much text between the end of "real text" and more articles + subscriber box, which isn't aesthetic.